tmol/pose/pose_stack.py

```diff
@@ -0,0 +1,716 @@
+import attr
+import toolz
+
+import itertools
+
+import numpy
+import torch
+import pandas
+import sparse
+import scipy.sparse.csgraph as csgraph
+
+from typing import Sequence, Tuple
+
+from tmol.types.array import NDArray
+from tmol.types.torch import Tensor
+
+from tmol.chemical.restypes import (
+    RefinedResidueType,
+    Residue,
+    find_simple_polymeric_connections,
+)
+from tmol.pose.packed_block_types import PackedBlockTypes, residue_types_from_residues
+from tmol.system.datatypes import connection_metadata_dtype
+from tmol.utility.tensor.common_operations import exclusive_cumsum1d
+from tmol.types.functional import validate_args
+
+
+@attr.s(auto_attribs=True)
+class PoseStack:
+
+    packed_block_types: PackedBlockTypes
+    residues: Sequence[Sequence[Residue]]
+
+    residue_coords: NDArray[numpy.float32][:, :, :, 3]
+
+    # coordinates are held as [n-poses x max-n-atoms x 3]
+    # where the offset for each residue are held in the
+    # block_coord_offset tensor [n-poses x max-n-blocks]
+    coords: Tensor[torch.float32][:, :, 3]
+
+    block_coord_offset: Tensor[torch.int32][:, :]
+    block_coord_offset64: Tensor[torch.int64][:, :]
```

Comment:
just curious, why the distinction?  is it because torch needs int64 for indexing?  Why not just use int64?

Reply:
I don't want to use int64 in our kernels because it's a waste of bandwidth and local memory / shared memory when we have no situations in which we will ever hit our heads on the 2 billion upper limit on the number of things represented by a signed integer. However, yes, torch wants 64 bit integers for all of their operations, and we still write a decent number of things using torch. So the choice is then either 1) keep int32 as a data member and convert to int64 only as often as its needed in a `block_coord_offset.to(torch.int64)` call or 2) keep both int32 and int64 around. It felt like the number of `.to` calls I needed was so high that it was wasteful to launch a kernel just to make one of these conversions and then to use the data a single time before throwing it away (only to re-create it again soon thereafter).